xnmt/hyper_parameters.py

```diff
@@ -2,91 +2,74 @@
 from xnmt.events import register_xnmt_handler, handle_xnmt_event
 from xnmt.persistence import serializable_init, Serializable
 
-class ScalingParam(Serializable):
-  """ initial * scaler(epoch-1) """
-  yaml_tag = "!ScalingParam"
-
-  @serializable_init
-  def __init__(self, initial=0.0, scaler=None):
-    self.__value = initial
-    self.scaler = scaler
-
-  def value(self):
-    value = self.__value
-    if self.scaler is not None:
-      value *= self.scaler.value()
-    return value
-
-  def __repr__(self):
-    return str(self.value())
-
 class Scalar(Serializable):
-  yaml_tag = "!Scalar"
+  """
+  Scalar class for hyper parameter that support 1 value serialization.
+  This class is actually a base class and does not have any different with simple python float/int.
+  
+  Args:
+    initial: The value being hold by the scalar.
+    update: Is the epoch number. 
+  """
 
+  yaml_tag = "!Scalar"
+  
   @serializable_init
-  def __init__(self, initial=0.0):
-    self.__value = initial
-
-  def value(self):
-    return self.__value
-
-  def __repr__(self):
-    return str(self.value())
-
-class GeometricSequence(Serializable):
-  """ initial^(epoch) """
-  yaml_tag = '!GeometricSequence'
-
-  # Do not set warmup_counter manually.
   @register_xnmt_handler
-  @serializable_init
-  def __init__(self, initial=0.1, warmup=0, ratio=1, min_value=0.0, max_value=1.0):
-    self.__value = initial
-    self.warmup = warmup
-    self.ratio = ratio
-    self.min_value = min_value
-    self.max_value = max_value
-    self.epoch_num = 0
-
-  def value(self):
-    if self.epoch_num >= self.warmup:
-      return self.__value
-    else:
-      return 0.0
+  def __init__(self, initial=0.0, update=0):
+    self.value = initial
+    self.update = update
 
   @handle_xnmt_event
-  def on_new_epoch(self, training_task, *args, **kwargs):
-    self.epoch_num = training_task.training_state.epoch_num
-    if self.epoch_num > self.warmup:
-      value = self.__value * self.ratio
-      value = max(self.min_value, value)
-      value = min(self.max_value, value)
-      self.__value = value
-
-  def __repr__(self):
-    return repr(self.value())
+  def on_new_epoch(self, *args, **kwargs):
+    self.value = self.update_value()
+    self.update += 1
+    self.save_processed_arg("initial", self.value)
+    self.save_processed_arg("update", self.update)
+ 
+  def update_value(self):
+    return self.value
+
+  # Operators
+  def __lt__(a, b): return a.value < b
+  def __le__(a, b): return a.value <= b
+  def __eq__(a, b): return a.value == b
+  def __ne__(a, b): return a.value != b
+  def __ge__(a, b): return a.value >= b
+  def __gt__(a, b): return a.value > b
```

Comment:
This is minor but I think it's possible to simplify this a bit using ```@functools.total_ordering```

Reply:
I took a look at the implementation and I think it's a bit overhead to do so. It is also easier to understand by listing this. So let's keep it this way?

Reply:
Sure, I think that's perfectly fine.